system-addon/content-src/components/TopSites/TopSite.jsx

```diff
@@ -8,59 +8,125 @@ import {
 import {LinkMenu} from "content-src/components/LinkMenu/LinkMenu";
 import React from "react";
 
-export const TopSiteLink = props => {
-  const {link, title} = props;
-  const topSiteOuterClassName = `top-site-outer${props.className ? ` ${props.className}` : ""}`;
-  const {tippyTopIcon, faviconSize} = link;
-  const letterFallback = title[0];
-  let imageClassName;
-  let imageStyle;
-  let showSmallFavicon = false;
-  let smallFaviconStyle;
-  let smallFaviconFallback;
-  if (tippyTopIcon || faviconSize >= MIN_RICH_FAVICON_SIZE) {
-    // styles and class names for top sites with rich icons
-    imageClassName = "top-site-icon rich-icon";
-    imageStyle = {
-      backgroundColor: link.backgroundColor,
-      backgroundImage: `url(${tippyTopIcon || link.favicon})`
-    };
-  } else {
-    // styles and class names for top sites with screenshot + small icon in top left corner
-    imageClassName = `screenshot${link.screenshot ? " active" : ""}`;
-    imageStyle = {backgroundImage: link.screenshot ? `url(${link.screenshot})` : "none"};
+export class TopSiteLink extends React.PureComponent {
+  constructor(props) {
+    super(props);
+    this.onDragStart = this.onDragStart.bind(this);
+    this.onDragEnd = this.onDragEnd.bind(this);
+    this.onDragOver = this.onDragOver.bind(this);
+    this.onDragEnter = this.onDragEnter.bind(this);
+    this.onDragLeave = this.onDragLeave.bind(this);
+    this.onDrop = this.onDrop.bind(this);
```

Comment:
We have this pattern in many places now, but I wonder if having a central handler to dispatch would be any nicer. E.g.,
```js
constructor() {
  this.onEvent = this.onEvent.bind(this);
}
onEvent(event) {
  switch (event.type) {
    case "drop": …
}
```

Eh.. probably too much :p Maybe I'm thinking more of the `handleEvent` method on the `EventListener` interface, so one could just do `addEventListener("drop", this);` but I don't think React handles that? https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener

Reply:
yeah, I’m definitely “tired” of the .bind() boilerplate